shell/plugins/dashboard-store/mutations.js

```diff
@@ -165,6 +180,71 @@ export function remove(state, obj, getters) {
   }
 }
 
+export function batchChanges(state, { ctx, batch }) {
+  const types = Object.keys(batch);
+
+  types.forEach((type) => {
+    const normalizedType = normalizeType(type === 'counts' ? COUNT : type);
+    const keyField = keyFieldFor(normalizedType);
+    const typeCache = registerType(state, normalizedType);
+    const opts = ctx.rootGetters[`type-map/optionsFor`](type);
+    const limit = opts?.limit;
+
+    const createdResources = [];
+    const batchActionMap = Object.keys(batch[normalizedType]).map((id) => {
+      const index = typeCache.list.findIndex((resource) => {
+        return resource[keyField] === id;
+      });
+      const resource = batch[normalizedType][id];
+      let action = 'change';
+
+      if (resource.remove) {
+        action = 'remove';
+      } else if (index === -1) {
+        action = 'create';
+        createdResources.push(specialFields(normalizedType, resource));
+      }
+
+      return {
+        id,
+        index,
+        resource: action === 'change' ? specialFields(normalizedType, resource) : undefined,
+        action
+      };
+    }).reduce((acc, batchIdMap) => {
+      return {
+        ...acc,
+        [batchIdMap[keyField]]: batchIdMap
+      };
+    }, {});
+
+    const newCache = typeCache.list.reduce((acc, resource) => {
```

Comment:
I did go ahead and move the map add/setting into the reduce function instead of looping over the array again.

If the limit is lower than the count the slice is a simple no-op and shouldn't impact the complexity of the mutation at all

You're right, I was using addObjects just because that's what we were doing in other mutations but it's more than I need here so I swapped it out for a simple push.

I could further refine this, by using the index on the cache to directly modify indexes, add new resources, and then slice out the removed ones but I don't think we'd be getting much by cutting out a linear loop over the array (at the end of the day, simple loops over an array that's only in the 10s of thousands is in the 10s of milliseconds) and done is better than perfect here probably.